File: ios/opensleep/opensleep/Views/Components/SleepCharts.swift
import SwiftUI

// MARK: - Donut Ring Chart

struct SleepRingChart: View {
    let stageDurations: [SleepStageType: TimeInterval]
    @AppStorage("app_language") private var selectedLanguage = "en"

    private var languageBundle: Bundle {
        if let path = Bundle.main.path(forResource: selectedLanguage, ofType: "lproj"),
           let bundle = Bundle(path: path) {
            return bundle
        }
        return Bundle.main
    }

    private func localizedString(_ key: String) -> String {
        return languageBundle.localizedString(forKey: key, value: nil, table: nil)
    }

    private var total: TimeInterval {
        stageDurations.values.reduce(0, +)
    }

    private var nonZero: [(SleepStageType, TimeInterval)] {
        stageDurations.filter { $0.value > 0 }
            .sorted { $0.key.rawValue < $1.key.rawValue }
    }

    var body: some View {
        ZStack {
            Canvas { ctx, size in
                let diameter = min(size.width, size.height)
                let lineWidth: CGFloat = 24
                var startAngle = Angle.degrees(-90)
                for (type, duration) in nonZero {
                    let fraction = duration / total
                    let sweep = Angle.degrees(fraction * 360)
                    var path = Path()
                    path.addArc(center: CGPoint(x: size.width / 2, y: size.height / 2),
                               radius: (diameter - lineWidth) / 2,
                               startAngle: startAngle,
                               endAngle: startAngle + sweep,
                               clockwise: false)
                    ctx.stroke(path,
                               with: .color(Color.forStage(type)),
                               style: StrokeStyle(lineWidth: lineWidth, lineCap: .butt))
                    startAngle += sweep
                }
            }

            // Center label
            VStack(spacing: 2) {
                let h = Int(total / 3600)
                let m = Int((total.truncatingRemainder(dividingBy: 3600)) / 60)
                Text("\(h)h \(m)m")
                    .font(.system(size: 18, weight: .bold, design: .rounded))
                    .foregroundStyle(Color.textPrimary)
                Text(localizedString("total"))
                    .font(AppTextStyle.labelSmall)
                    .foregroundStyle(Color.textSecondary)
            }
        }
    }
}

// MARK: - Hypnogram

struct SleepHypnogram: View {
    let stages: [SleepStage]
    var sessionStartDate: Date? = nil
    let totalDuration: TimeInterval

    private let stageOrder: [SleepStageType] = [.awake, .rem, .light, .deep]

    private var startDateTime: Date {
        sessionStartDate ?? stages.first?.startDate ?? Date()
    }
    
    private var endDateTime: Date {
        startDateTime.addingTimeInterval(totalDuration)
    }

    var body: some View {
        VStack(spacing: 8) {
            Canvas { ctx, size in
                guard !stages.isEmpty, totalDuration > 0 else { return }
                let trackH = size.height / CGFloat(stageOrder.count)
                let startTime = startDateTime.timeIntervalSinceReferenceDate

                // Draw background grid lines (25%, 50%, 75% marks)
                let gridLines = 3
                for i in 1...gridLines {
                    let fraction = CGFloat(i) / CGFloat(gridLines + 1)
                    let gx = fraction * size.width
                    var path = Path()
                    path.move(to: CGPoint(x: gx, y: 0))
                    path.addLine(to: CGPoint(x: gx, y: size.height))
                    ctx.stroke(path, with: .color(Color.textTertiary.opacity(0.15)), style: StrokeStyle(lineWidth: 1, dash: [2, 4]))
                }

                for stage in stages {
                    let x = CGFloat((stage.startDate.timeIntervalSinceReferenceDate - startTime) / totalDuration) * size.width
                    let w = max(CGFloat(stage.durationSeconds / totalDuration) * size.width, 2)
                    let yIdx = stageOrder.firstIndex(of: stage.type) ?? 0
                    let y = CGFloat(yIdx) * trackH

                    let rect = CGRect(x: x, y: y, width: w, height: trackH - 2)
                    ctx.fill(Path(roundedRect: rect, cornerRadius: 2),
                            with: .color(Color.forStage(stage.type).opacity(0.85)))
                }
            }
            .frame(height: 80)

            // Time Axis Labels
            HStack {
                Text(startDateTime.formatted(date: .omitted, time: .shortened))
                    .font(AppTextStyle.labelSmall)
                    .foregroundStyle(Color.textSecondary)
                
                Spacer()
                
                let midDate = startDateTime.addingTimeInterval(totalDuration / 2)
                Text(midDate.formatted(date: .omitted, time: .shortened))
                    .font(AppTextStyle.labelSmall)
                    .foregroundStyle(Color.textTertiary)
                
                Spacer()
                
                Text(endDateTime.formatted(date: .omitted, time: .shortened))
                    .font(AppTextStyle.labelSmall)
                    .foregroundStyle(Color.textSecondary)
            }
            .padding(.horizontal, 4)
        }
    }
}

// MARK: - Stage Legend

struct SleepStageLegend: View {
    let stageDurations: [SleepStageType: TimeInterval]

    var body: some View {
        HStack {
            ForEach(SleepStageType.allCases, id: \.self) { type in
                let minutes = Int((stageDurations[type] ?? 0) / 60)
                VStack(spacing: 4) {
                    Circle()
                        .fill(Color.forStage(type))
                        .frame(width: 8, height: 8)
                    Text(type.displayName)
                        .font(AppTextStyle.labelSmall)
                        .foregroundStyle(Color.forStage(type))
                    Text("\(minutes)m")
                        .font(AppTextStyle.labelSmall)
                        .foregroundStyle(Color.textSecondary)
                }
                .frame(maxWidth: .infinity)
            }
        }
    }
}

File: android/app/src/main/java/app/opensleep/ui/components/SleepCharts.kt
package app.opensleep.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.res.stringResource
import app.opensleep.R
import app.opensleep.data.local.SleepStageType
import app.opensleep.ui.theme.*
import java.util.concurrent.TimeUnit

fun stageNameRes(type: SleepStageType): Int = when (type) {
    SleepStageType.AWAKE -> R.string.stage_awake
    SleepStageType.LIGHT -> R.string.stage_light
    SleepStageType.DEEP -> R.string.stage_deep
    SleepStageType.REM -> R.string.stage_rem
}

fun stageColor(type: SleepStageType): Color = when (type) {
    SleepStageType.AWAKE -> ColorAwake
    SleepStageType.LIGHT -> ColorLight
    SleepStageType.DEEP -> ColorDeep
    SleepStageType.REM -> ColorREM
}

/**
 * Donut chart showing proportion of each sleep stage.
 */
@Composable
fun SleepRingChart(
    stageDurations: Map<SleepStageType, Long>,
    modifier: Modifier = Modifier
) {
    val totalMs = stageDurations.values.sum().takeIf { it > 0 } ?: return
    val nonZero = stageDurations.filter { it.value > 0 }
    val angles = nonZero.map { (type, ms) ->
        type to (ms.toFloat() / totalMs * 360f)
    }

    val strokeWidth = 28.dp

    Box(modifier = modifier, contentAlignment = Alignment.Center) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            val diameter = minOf(size.width, size.height)
            val strokePx = strokeWidth.toPx()
            val topLeft = Offset(
                (size.width - diameter) / 2 + strokePx / 2,
                (size.height - diameter) / 2 + strokePx / 2
            )
            val arcSize = Size(diameter - strokePx, diameter - strokePx)

            var startAngle = -90f
            angles.forEach { (type, sweep) ->
                drawArc(
                    color = stageColor(type),
                    startAngle = startAngle,
                    sweepAngle = sweep,
                    useCenter = false,
                    topLeft = topLeft,
                    size = arcSize,
                    style = Stroke(width = strokePx, cap = androidx.compose.ui.graphics.StrokeCap.Butt)
                )
                startAngle += sweep
            }
        }

        // Center text — total sleep hours
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            val hours = TimeUnit.MILLISECONDS.toHours(totalMs)
            val minutes = TimeUnit.MILLISECONDS.toMinutes(totalMs) % 60
            Text(
                text = "${hours}h ${minutes}m",
                style = MaterialTheme.typography.titleLarge.copy(
                    fontWeight = FontWeight.Bold,
                    fontSize = 20.sp,
                    color = TextPrimary
                )
            )
            Text(
                text = stringResource(R.string.total_sleep),
                style = MaterialTheme.typography.labelSmall.copy(color = TextSecondary)
            )
        }
    }
}

/**
 * Horizontal hypnogram — time-series strip chart showing sleep stages over the night.
 */
@Composable
fun SleepHypnogram(
    stages: List<app.opensleep.data.local.SleepStage>,
    sessionStartMs: Long? = null,
    totalDurationMs: Long,
    modifier: Modifier = Modifier
) {
    if (stages.isEmpty() || totalDurationMs <= 0) return

    val stageOrder = listOf(
        SleepStageType.AWAKE to 0f,
        SleepStageType.REM to 0.33f,
        SleepStageType.LIGHT to 0.66f,
        SleepStageType.DEEP to 1f
    )

    val finalStartMs = sessionStartMs ?: stages.firstOrNull()?.startMs ?: System.currentTimeMillis()
    val finalEndMs = finalStartMs + totalDurationMs

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Canvas(modifier = Modifier.height(80.dp).fillMaxWidth()) {
            val totalWidth = size.width
            val height = size.height
            val trackHeight = height / stageOrder.size

            // Draw background dotted vertical grid lines (25%, 50%, 75% marks)
            val gridLines = 3
            for (i in 1..gridLines) {
                val fraction = i.toFloat() / (gridLines + 1)
                val gx = fraction * totalWidth
                drawLine(
                    color = TextTertiary.copy(alpha = 0.15f),
                    start = Offset(gx, 0f),
                    end = Offset(gx, height),
                    strokeWidth = 1.dp.toPx(),
                    pathEffect = androidx.compose.ui.graphics.PathEffect.dashPathEffect(floatArrayOf(5f, 10f), 0f)
                )
            }

            stages.forEach { stage ->
                val x = ((stage.startMs - finalStartMs).toFloat() / totalDurationMs) * totalWidth
                val w = ((stage.endMs - stage.startMs).toFloat() / totalDurationMs) * totalWidth
                val yIdx = stageOrder.indexOfFirst { it.first == stage.type }
                val y = yIdx * trackHeight

                drawRect(
                    color = stageColor(stage.type).copy(alpha = 0.85f),
                    topLeft = Offset(x, y),
                    size = Size(w.coerceAtLeast(2f), trackHeight - 2f)
                )
            }
        }

        // Time axis labels
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 4.dp),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            val sdf = java.text.SimpleDateFormat("h:mm a", java.util.Locale.getDefault())
            val startTimeStr = sdf.format(java.util.Date(finalStartMs)).lowercase()
            val midTimeStr = sdf.format(java.util.Date(finalStartMs + totalDurationMs / 2)).lowercase()
            val endTimeStr = sdf.format(java.util.Date(finalEndMs)).lowercase()

            Text(
                text = startTimeStr,
                style = MaterialTheme.typography.labelSmall,
                color = TextSecondary
            )
            Text(
                text = midTimeStr,
                style = MaterialTheme.typography.labelSmall,
                color = TextTertiary
            )
            Text(
                text = endTimeStr,
                style = MaterialTheme.typography.labelSmall,
                color = TextSecondary
            )
        }
    }
}

/**
 * Legend row for sleep stages.
 */
@Composable
fun SleepStageLegend(
    stageDurations: Map<SleepStageType, Long>,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceEvenly
    ) {
        SleepStageType.entries.forEach { type ->
            val ms = stageDurations[type] ?: 0L
            val minutes = TimeUnit.MILLISECONDS.toMinutes(ms)
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Box(
                    modifier = Modifier
                        .size(10.dp)
                        .padding(bottom = 2.dp)
                ) {
                    Canvas(modifier = Modifier.fillMaxSize()) {
                        drawCircle(color = stageColor(type))
                    }
                }
                Spacer(Modifier.height(4.dp))
                Text(
                    text = stringResource(stageNameRes(type)),
                    style = MaterialTheme.typography.labelSmall,
                    color = stageColor(type)
                )
                Text(
                    text = "${minutes}m",
                    style = MaterialTheme.typography.labelSmall,
                    color = TextSecondary
                )
            }
        }
    }
}
